File: day_2/part_2.py
FILE_NAME = "input.txt"

def generate_dirs():
    dirs = []
    with open(FILE_NAME, 'r') as f:
        for line in f.readlines():
            dir, val = line.split(' ')
            val = int(val)
            dirs.append((dir, val))
    return dirs


def f():
    vertical = 0
    horizontal = 0
    aim = 0
    dirs = generate_dirs()

    for dir, val in dirs:
        if dir == 'forward':
            horizontal += val
            vertical += aim * val
        elif dir == 'up':
            aim -= val
        elif dir == 'down':
            aim += val

    return horizontal * vertical

if __name__ == "__main__":
    print(f())

        

File: day_02/src/part_2.rs
use std::include_str;

struct Thing {
    direction: String,
    magnitude: i32,
}

fn main() {
    let ns = include_str!("../input.txt")
        .lines()
        .map(|line| {
            let mut split = line.split(" ");
            let direction = split.next().unwrap().to_string();
            let magnitude = split.next().unwrap().parse::<i32>().unwrap();

            Thing { direction, magnitude }
        });

    let mut horizontal: i32 = 0;
    let mut vertical: i32 = 0;
    let mut aim: i32 = 0;

    for n in ns {
        if n.direction == "up" {
            aim -= n.magnitude;
        }
        else if n.direction == "down" {
            aim += n.magnitude;
        }
        else { // n.direction == "forward"
            horizontal += n.magnitude;
            vertical += aim * n.magnitude;
        }
    }

    println!("{}", horizontal * vertical);
}
